lib/core/network/protocols/http+websocket.js

```diff
@@ -0,0 +1,920 @@
+        return;
+      }
+
+      this.httpProcessRequest(response, message);
+    });
+  }
+
+  /**
+   * Read an HTTP payload data
+   *
+   * @param  {HttpMessage} message
+   * @param  {uWS.HttpResponse} response
+   * @param  {Function} cb
+   */
+  httpReadData (message, response, cb) {
+    let payload = null;
+    response.aborted = false;
+
+    response.onData((data, isLast) => {
+      const chunk = Buffer.from(data);
+
+      payload = payload
+        ? Buffer.concat([payload, chunk])
+        : Buffer.concat([chunk]);
+
+      /*
+       * The content-length header can be bypassed and
+       * is not reliable enough. We have to enforce the HTTP
+       * max size limit while reading chunks too
+       */
+      if (payload.byteLength > this.maxRequestSize) {
+        cb(HTTP_REQUEST_TOO_LARGE_ERROR);
+        return;
+      }
+
+      if (!isLast) {
+        return;
+      }
+
+      if (payload.byteLength === 0) {
+        cb();
+        return;
+      }
+
+      this.httpUncompress(message, payload, (err, inflated) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+
+        this.httpParseContent(message, inflated, cb);
+      });
+    });
+
+    // Beware: the "response" object might be invalidated at any point of time,
+    // whenever the connection gets closed by the client for instance
+    response.onAborted(() => {
+      response.aborted = true;
+    });
+  }
+
+  httpParseContent (message, content, cb) {
+    const type = message.headers['content-type'] || '';
+
+    if (type.includes('multipart/form-data')) {
+      const parts = uWS.getParts(content, message.headers['content-type']);
+      message.content = {};
+
+      for (const part of parts) {
+        if (part.data.byteLength > this.maxFormFileSize) {
+          cb(HTTP_FILE_TOO_LARGE_ERROR);
+        }
+
+        if (part.filename) {
+          message.content[part.name] = {
+            encoding: part.type,
+            file: Buffer.from(part.data).toString('base64'),
+            filename: part.filename,
+          };
+        }
+        else {
+          message.content[part.name] = Buffer.from(part.data).toString();
+        }
+      }
+    }
+    else if (type.includes('application/x-www-form-urlencoded')) {
+      message.content = querystring.parse(content.toString());
+    }
+    else {
+      try {
+        message.content = JSON.parse(content.toString());
+      }
+      catch (e) {
+        cb(kerrorHTTP.get('body_parse_failed', content.toString().slice(0, 50)));
+        return;
+      }
+    }
+
+    cb();
+  }
+
+  httpProcessRequest (response, message) {
+    debugHTTP('[%s] httpProcessRequest: %a', message.connection.id, message);
+
+    if (response.aborted) {
+      return;
+    }
+
+    this.entryPoint.newConnection(message.connection);
+
+    global.kuzzle.router.http.route(message, request => {
+      this.entryPoint.logAccess(request, message);
+      this.entryPoint.removeConnection(message.connection.id);
```

Comment:
Why removing the connection at the beginning and not at the end?

Reply:
Because there are multiple exit routes: the connection might have already been aborted, or it can aborts before we finish compressing the response, or during the time we send the response (see tryEnd/onWritable)

We might as well remove the ClientConnection object now: it's only the Kuzzle representation of a network connection, and at that point of time, it doesn't exist anymore (its associated request has been processed already)